pkg/route/hostassignment/externalcertificate.go

```diff
@@ -1,35 +0,0 @@
-package hostassignment
-
-import (
-	"context"
-
-	"k8s.io/apimachinery/pkg/util/validation/field"
-
-	routev1 "github.com/openshift/api/route/v1"
-	routecommon "github.com/openshift/library-go/pkg/route"
-)
-
-// ValidateHostExternalCertificate checks if the user has permissions to create and update
-// custom-host subresource of routes. This check is required to be done prior to ValidateHostUpdate()
-// since updating hosts while using externalCertificate is contingent on the user having both these
-// permissions. The ValidateHostUpdate() cannot differentiate if the certificate has changed since
-// now the certificates will be present as a secret object, due to this it proceeds with the assumption
-// that the certificate has changed when the route has externalCertificate set.
-// TODO: Consider merging this function into ValidateHostUpdate.
-func ValidateHostExternalCertificate(ctx context.Context, new, older *routev1.Route, sarc routecommon.SubjectAccessReviewCreator, opts routecommon.RouteValidationOptions) field.ErrorList {
-
-	if !opts.AllowExternalCertificates {
-		// Return nil since the feature gate is off.
-		// ValidateHostUpdate() is sufficient to validate
-		// permissions.
-		return nil
-	}
-
-	newTLS := new.Spec.TLS
-	oldTLS := older.Spec.TLS
-	if (newTLS != nil && newTLS.ExternalCertificate != nil) || (oldTLS != nil && oldTLS.ExternalCertificate != nil) {
-		return routecommon.CheckRouteCustomHostSAR(ctx, field.NewPath("spec", "tls", "externalCertificate"), sarc)
-	}
-
-	return nil
-}
```

Comment:
Since `ValidateHostExternalCertificate` is used in openshift-apiserver, removing its definition now could cause difficulties if someone else needs a change from library-go in openshift-apiserver and tries to bump it before you update openshift-apiserver not to use `ValidateHostExternalCertificate`.  That is:

1. You remove `ValidateHostExternalCertificate` from library-go.  
2. You don't have time to post the openshift-apiserver PR right away, or reviews get delayed.  
3. Meanwhile, someone else adds function xyz to library-go and bumps openshift-apiserver to get xyz.
4. Building openshift-apiserver fails because `ValidateHostExternalCertificate` is not defined.  

It would be safer to do the following:

1. Update logic in library-go, but keep `ValidateHostExternalCertificate` (maybe add a `// Deprecated` comment) so library-go can be bumped in openshift-apiserver without breaking anything.  
5. Update code in openshift-apiserver not to use `ValidateHostExternalCertificate`.
6. Remove `ValidateHostExternalCertificate` from library-go.

Reply:
Thanks for the suggestion, and I agree that we should be doing the way you suggested above. 

However, even if we keep `ValidateHostExternalCertificate` definition, the [unit tests present in the openshift-apiserver](https://github.com/openshift/openshift-apiserver/blob/cb676917bae9003e14277684f1606d12b09bf484/pkg/route/apiserver/registry/route/strategy_test.go#L153) will complain due to the changes in `validateTLSExternalCertificate` function because we are dropping `custom-host` permission check from that function as well.  There are fair amount of unit tests overlap between `openshift-apiserver` and library-go, so updating logic in library-go requires immediate intervention (bump) in `openshift-apiserver` as well. 

If these weren't any unit tests in `openshift-apiserver`, then it could have been much easier.